react/src/utils/Client.js

```diff
@@ -0,0 +1,143 @@
+    if (!Str.HasValue(path)) {
+        path = GetCurrentLogicalPath();
+    }
+    if (!path.startsWith("/")) {
+        path = "/" + path;
+    }
+    path = Path.Normalize(path);
+    if (Type.IsBoolean(env)) {
+        if (env) {
+            //
+            // Use the current environment (from the current URL) for the given path.
+            // If it doesn't exist (for some reason) use the default environemnet
+            // from the global header data, if passed in.
+            //
+            env = Env.Current();
+            if (!Str.HasValue(env)) {
+                if (Str.HasValue(envFallback)) {
+                    env = envFallback;
+                }
+                else if (Type.IsObject(envFallback)) {
+                    const header = envFallback;
+                    // env = header.envs?.default;
+                    env = Env.Default(header);
+                }
+            }
+        }
+
+    }
+    else if (Type.IsObject(env)) {
+        const header = env;
+        // env = header.envs?.default;
+        env = Env.Default(header);
+        if (!Str.HasValue(env)) {
+            if (Str.HasValue(envFallback)) {
+                env = envFallback;
+            }
+            else if (Type.IsObject(envFallback)) {
+                //
+                // This would be weird but just for completeness.
+                //
+                const header = envFallback;
+                // env = header.envs?.default;
+                env = Env.Default(header);
+            }
+        }
+    }
+    if (Str.HasValue(env)) {
+        path = "/" + env + path;
+    }
+    path = Context.Client.BasePath() + path;
+    if (path.endsWith("/")) {
+        path = path.substring(0, path.length - 1);
+    }
+    return path;
+}
+
+// -------------------------------------------------------------------------------------------------
+// Current client path functions.
+// -------------------------------------------------------------------------------------------------
+
+function GetCurrentLogicalPath() {
+    const currentPath = Path.Normalize(window.location.pathname);
+    const basePathWithTrailingSlash = Context.Client.BasePath() + "/";
+    if (currentPath.startsWith(basePathWithTrailingSlash)) {
+        const pathSansBasePath = currentPath.substring(basePathWithTrailingSlash.length);
+        if (pathSansBasePath.length > 0) {
+            const slash = pathSansBasePath.indexOf("/");
+            if (slash >= 0) {
+                return pathSansBasePath.substring(slash);
+            }
+        }
+    }
+    return "";
+}
+
+// -------------------------------------------------------------------------------------------------
+// Links functions.
+// -------------------------------------------------------------------------------------------------
+
+function GetPortalLink(header) {
+    if (Env.IsFoursightFourfront(header)) {
+        return "https://" +  Env.PublicName(Env.Current(), header) + ".4dnucleome.org/";
```

Comment:
I don't think this will always work, it's probably better to just extract the URL directly from the env bucket.

Reply:
Per related comment above added TODO comment to code.